src/com/ms/silverking/pssh/TwoLevelParallelSSHMaster.java

```diff
@@ -34,460 +34,443 @@
+  public void workerComplete() throws RemoteException {
+    try {
+      Log.warning("worker complete  from " + getClientHost());
+    } catch (ServerNotActiveException snae) {
+      snae.printStackTrace();
     }
+    completedWorkers.release();
+    Log.warning("Worker complete: ", completedWorkers.availablePermits());
+  }
+
+  //////////////////////////////////////////////////////////////////////
 
-    private void logState() {
-        Log.warning("Complete: "+ completeHostCommands.size()
-                 +"\tIncomplete: "+ incompleteHostCommands.size() 
-                 +"\tActive: "+ activeHostCommands.size() 
-                 +"\tPending: "+ pendingHostCommands.size());
+  public void displayIncomplete() {
+    for (HostAndCommand hostAndCommand : incompleteHostCommands) {
+      System.out.println("Incomplete:\t" + hostAndCommand);
     }
-    
-    @Override
-    public void setHostResult(HostAndCommand hostAndCommand, HostResult result) {
-        try {
-            Log.warning("set host result "+ hostAndCommand +" from "+ getClientHost());
-        } catch (ServerNotActiveException snae) {
-            snae.printStackTrace();
-        }
-        activeHostCommands.remove(hostAndCommand);
-        incompleteHostCommands.remove(hostAndCommand);
-        completeHostCommands.add(hostAndCommand);
-        if (incompleteHostCommands.size() == 0) {
-            try {
-                pendingHostCommands.put(DONE_MARKER);
-            } catch (InterruptedException ie) {
-            }
-            if (terminateUponCompletion) {
-                if (beginTermination()) {
-                    displayIncomplete();
-                }
-            } else {
-                displayIncomplete();
-            }
-        }
-        logState();
+  }
+
+  public static boolean beginTermination() {
+    if (!terminated.getAndSet(true)) {
+      new com.ms.silverking.process.Terminator(terminationSeconds, 10);
+      return true;
     }
+    return false;
+  }
 
-    @Override
-    public void workerComplete() throws RemoteException {
-        try {
-            Log.warning("worker complete  from "+ getClientHost());
-        } catch (ServerNotActiveException snae) {
-            snae.printStackTrace();
-        }
-        completedWorkers.release();
-        Log.warning("Worker complete: ", completedWorkers.availablePermits());
-    }    
-        
-    //////////////////////////////////////////////////////////////////////
-    
-    public void displayIncomplete() {
-        for (HostAndCommand hostAndCommand : incompleteHostCommands) {
-            System.out.println("Incomplete:\t"+ hostAndCommand);
-        }
+  public static List<String> readHostsFile(String hostsFile, String description, List<String> exclusionHosts)
+      throws IOException {
+    List<String> hostList;
+
+    if (hostsFile != null) {
+      hostList = StreamParser.parseFileLines(hostsFile);
+    } else {
+      Log.warning("No " + description + " file specified");
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;description&#39; (index: 1 | type: String). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;hostsFile&#39;, &#39;exclusionHosts&#39;. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, reduce the method&#39;s visibility.

Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).